Invalid suggestion for accessing fields on raw pointer in macros

### Code

```Rust
struct Demo {
    val: u32,
}

macro_rules! as_ptr {
    ($d:expr) => {
        &mut $d as *mut Demo
    };
}

macro_rules! get_value {
    ($d:expr) => {
        as_ptr!($d).val
    }
}

fn main() {
    let d = Demo { val: 123 };
    assert_eq!(123, get_value!(d));
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0609]: no field `val` on type `*mut Demo`
  --> src/main.rs:13:21
   |
13 |         as_ptr!($d).val
   |                     ^^^ unknown field
...
19 |     assert_eq!(123, get_value!(d));
   |                     ------------- in this macro invocation
   |
   = note: this error originates in the macro `get_value` (in Nightly builds, run with -Z macro-backtrace for more info)
help: `$d as *mut Demo` is a raw pointer; try dereferencing it
   |
 7 -         &mut $d as *mut Demo
 7 +         &mut (*).
   |

For more information about this error, try `rustc --explain E0609`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

### Desired output

```Shell

```

### Rationale and extra context

At the very least,
```
 7 -         &mut $d as *mut Demo
 7 +         &mut (*).
```
is broken, if a well-formed suggestion cannot be provided it should be removed

### Other cases

With


```Rust
struct Demo {
    val: u32,
}

macro_rules! get_value {
    ($d:expr) => {
        $d.val
    }
}

fn main() {
    let d = Demo { val: 123 };
    let d_ptr = &d as *const Demo;
    assert_eq!(123, get_value!(d_ptr));
}
```

the output is still problematic, but less so since it doesn't have a suggestion to apply:

```
   Compiling playground v0.0.1 (/playground)
error[E0609]: no field `val` on type `*const Demo`
  --> src/main.rs:7:12
   |
 7 |         $d.val
   |            ^^^ unknown field
...
14 |     assert_eq!(123, get_value!(d_ptr));
   |                     ----------------- in this macro invocation
   |
   = note: this error originates in the macro `get_value` (in Nightly builds, run with -Z macro-backtrace for more info)
help: `d_ptr` is a raw pointer; try dereferencing it
   |
 7 ~         $d).val
 8 |     }
...
13 |     let d_ptr = &d as *const Demo;
14 ~     assert_eq!(123, get_value!((*d_ptr));
   |

For more information about this error, try `rustc --explain E0609`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

### Rust Version

```Shell
1.96.0
```

### Anything else?

_No response_

--- a/compiler/rustc_hir_typeck/src/expr.rs
+++ b/compiler/rustc_hir_typeck/src/expr.rs
@@ -3159,6 +3159,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
     fn suggest_first_deref_field(&self, err: &mut Diag<'_>, base: &hir::Expr<'_>, field: Ident) {
         err.span_label(field.span, "unknown field");
+        if base.span.from_expansion() || field.span.from_expansion() {
+            return;
+        }
         let val = if let Ok(base) = self.tcx.sess.source_map().span_to_snippet(base.span)
             && base.len() < 20
         {

--- a/tests/ui/macros/deref-raw-pointer-issue-158158.rs
+++ b/tests/ui/macros/deref-raw-pointer-issue-158158.rs
@@ -1,0 +1,21 @@
+struct Demo {
+    val: u32,
+}
+
+macro_rules! as_ptr {
+    ($d:expr) => {
+        &mut $d as *mut Demo
+    };
+}
+
+macro_rules! get_value {
+    ($d:expr) => {
+        as_ptr!($d).val
+        //~^ ERROR no field `val` on type `*mut Demo`
+    }
+}
+
+fn main() {
+    let mut d = Demo { val: 123 };
+    let _ = get_value!(d);
+}

--- a/tests/ui/macros/deref-raw-pointer-issue-158158.stderr
+++ b/tests/ui/macros/deref-raw-pointer-issue-158158.stderr
@@ -8,11 +8,6 @@ LL |     let _ = get_value!(d);
    |             ------------- in this macro invocation
    |
    = note: this error originates in the macro `get_value` (in Nightly builds, run with -Z macro-backtrace for more info)
-help: `$d as *mut Demo` is a raw pointer; try dereferencing it
-   |
-LL -         &mut $d as *mut Demo
-LL +         &mut (*).
-   |
 
 error: aborting due to 1 previous error
 

--- a/tests/ui/macros/deref-raw-pointer-issue-158158.stderr
+++ b/tests/ui/macros/deref-raw-pointer-issue-158158.stderr
@@ -1,0 +1,19 @@
+error[E0609]: no field `val` on type `*mut Demo`
+  --> $DIR/deref-raw-pointer-issue-158158.rs:13:21
+   |
+LL |         as_ptr!($d).val
+   |                     ^^^ unknown field
+...
+LL |     let _ = get_value!(d);
+   |             ------------- in this macro invocation
+   |
+   = note: this error originates in the macro `get_value` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: `$d as *mut Demo` is a raw pointer; try dereferencing it
+   |
+LL -         &mut $d as *mut Demo
+LL +         &mut (*).
+   |
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0609`.

